### PR TITLE
Second Draft of PM Process

### LINKED ISSUE #165: Provide a clear definition of "Backlog" with guidelines for using it
There was a fair about of discussion about the meaning of a "Backlog" in the PM pipeline and even more discussion about how to name sections within the backlog -- future work, icebox, ready, etc. I've gathered all of that info here.
## Work to be Done

Update the PM document to provide a clear definition of "backlog" with 
- a clear definition of the term backlog
- something that explains why this concept is important and declares how to be "consistent" with this model's usage of that term
- an example accompanied with a diagram
- guidelines for using the backlog
- suggestions for sections within the backlog and what to name them
## Confused by the Definition

I'm confused by the definition currently in the PM Doc:

> Backlog is the collection of all goals in a project. [...]

By that definition, the Backlog includes Goals that are In Progress and Done as well as the Future Work, Ready, etc. This doesn't fit with the way I see the term used in the rest of the document.

The document later uses the term "Backlog" to label both a _phase_ and a _stage_ in the same diagram, but the document doesn't define the term phase or explain how the notion of phases should be used to arrange Goals.
## Example from go-ipfs

In applying this model to go-ipfs (see [notes](https://github.com/ipfs/pm/blob/master/meeting-notes/2016-08-30-go-ipfs-pm-magic.md)) @whyrusleeping pointed out that most projects will need to communicate where each Goal lies on the spectrum shown in the table below. Based on our discussion, we came up with some ideas about how we can use pipeline statuses and Milestones to express this info. I included that info in the table. _The specific choice of names for these statuses is changeable, but the key ideas are important._

| What you want to communicate | Status | Milestones |
| --- | --- | --- |
| "This Goal was reported but maintainers haven't looked at it yet" | New/Inbox | No milestone or milestone without a date |
| "This Goal has been reviewed but needs more attention before it will be ready to work on" | Refining | Might have a milestone |
| "This Goal is well defined but not planned to be carried out at this moment. (ie. it doesn't have a priority or a deadline)" | Backlog | No milestone or milestone without a date |
| "This Goal has been reviewed and is ready to work on" | Ready | Might have a milestone |
| "This Goal is ready to work on and we will be working on it soon" | Ready | Has a milestone with a date |

This has a lot of overlap with these definitions in the draft doc:

> Future Work - the goal is identified and being fleshed out. It is not ready to be carried out at this moment.
> Backlog - the goal is well defined but not planned to be carried out at this moment.
> Ready - the goal is planned to be carried out, waiting to be worked on.

... but note that we reconsidered the terms "Future Work" and "Backlog" in [these comments](https://github.com/ipfs/pm/pull/131/files/0fbf77efecf51695f4c9370c540e4d6727cda80d#r74540072)
## What is the goal here?

What are the consistencies we are trying to achieve with the term "Backlog"? If we're establishing this model in order to have consistency of language and core structures across our projects, what are we asking people to be consistent about?

This diagram from the draft PM document presents "Backlog" as both a _stage_ and a _phase_, where the Backlog phase includes the Backlog stage and some other stages. **Is that intentional? Is that going to confuse people?** If we just need a different name for the phase, someone taking a scrum approach might label the "Backlog" phase "Refining" or "Ideation". 

> ```
>   ╭─────────────────────────────────── Pipeline ──────────────────────────────────────╮
>   │────── 1. ────│──── 2. ───│─── 3. ──│────── 4. ─────│─── 5. ───│──── 6. ───│── 7. ─│
>   │ Stages                                                                            │
>                                                 ╭────────────╮
> ──> Future Work ──> Backlog ──> Ready ──> In Progress ──> Review ──> Release ──> Done ──>
>          │             │          │             │            │ 
>          ╰─────────────╯          │             │            │
>                                   ╰───────── Blocked ────────╯
>     Phases
>   │───────── Backlog ────────│──────────── In Progress ───────────│──── Complete ─────│
>   ╰───────────────────────────────────────────────────────────────────────────────────╯
> ```

The document defines the stages with this key:

> **Future Work** - the goal is identified and being fleshed out. It is not ready to be carried out at this moment.
> **Backlog** - the goal is well defined but not planned to be carried out at this moment.
> **Ready** - the goal is planned to be carried out, waiting to be worked on.
> **In Progress** - the goal is actively being worked on, this moment.
> **Review** - the development of the goal is done, but a review needs to happen before it can be included in a release.
> **Release** - the goal is waiting to be inckuded in a an official release.
> **Done** - the goal is completed. No further action is necessary.
> **Blocked** - the goal cannot proceed as something is preventing it to move down the pipeline.

For go-ipfs, @whyrusleeping wants a slightly different pipeline:

```
  ╭─────────────────────────────────── Pipeline ──────────────────────────────────────╮
  │────── 1. ────│──── 2. ───│─── 3. ──│────── 4. ─────│─── 5. ───│──── 6. ───│── 7. ─│
  │ Stages                                                                            │

                                               ╭────────────╮
     ──> Inbox ──> Refining ──> Backlog ──> Ready ──> In Progress ──> Done ──>
                    │             │            │            │ 
                    ╰─────────────╯            │            │
                                            Blocked ────────╯
    Phases
  │────────────── Backlog ─────────────│────── In Progress ───────│──── Complete ─────│
  ╰───────────────────────────────────────────────────────────────────────────────────╯
```

This re-uses most of the stages but makes these additions/changes:
- _Future Work is not used_
- **Inbox** - the goal has not been reviewed or categorized yet
- **Refining** - the goal has been reviewed but is not ready to be worked on (ie. the description needs clarification, etc.) _This is similar to the Future Work stage in the previous diagram._

Important: In both of these diagrams the "Backlog" stage means the same thing. **Is that the point of consistency that we're aiming for?** 
## An attempt at Concise Statement

Is this an accurate representation of the consistency we're trying to achieve?

> All or our projects have pipeline stages for Backlog, Ready, In Progress and Done. All of those stages will have the same meaning across the projects. Some projects will have additional stages in their pipeline.

Alternatively, are we just saying:

> All of our project pipelines distinguish between the three phases of Backlog, In Progress and Done

What else are we trying to communicate?
## Background - References to Past Comments on this Topic

[These comments](https://github.com/ipfs/pm/pull/131/files/0fbf77efecf51695f4c9370c540e4d6727cda80d#r74540072) talk about this topic at length and include one [definition of Backlog](https://github.com/ipfs/pm/pull/131/files/0fbf77efecf51695f4c9370c540e4d6727cda80d#r74737162)

> I feel "Backlog" should be a term for the place where all goals are collected in.

The comments seem to converge on some recommended names for sections within the Backlog called either [inbox and ready](https://github.com/ipfs/pm/pull/131/files/0fbf77efecf51695f4c9370c540e4d6727cda80d#r74737162) or [TODO and Ready](https://github.com/ipfs/pm/pull/131/files/0fbf77efecf51695f4c9370c540e4d6727cda80d#r74737444)

### LINKED ISSUE #166: Frame Milestones as crossover point between the timeline view (Roadmap) and the pipeline view (Status Board) of your Goals
I propose thinking of Roadmaps and Status Boards as two views onto the same set of Goals.
- Every project has a Roadmap and a Status Board
- The Roadmap is primarily concerned with projected timeline. It arranges Goals by their Milestones, which have deadlines or estimated dates of delivery.
- The Status Board is primarily concerned with the current state of work. It arranges Goals based on their status within the pipeline.
- Milestones are the main point of navigation/connection between these two contexts.
### How this lets us Frame the Concepts

Roadmaps and Status Boards provide different views on the same set of Goals in order to answer different questions and to support distinct types of decisions:
- If you want to know the current state of work, go to the Status Board and look at the Goals there.
- If you want to know about timelines involved in the project and estimated dates of delivery, go to the Roadmap and look at the Milestones there.

Milestones are the bridge between these two contexts:
- If you want to know the current state of work for a particular Milestone in the Roadmap, go to the board and look at that Milestone's Goals on the Board.
- If you're looking at a Goal and you want to know how it fits into the overall Roadmap and timeline, find the Milestone that the Goal belongs to and see where that Milestone lies within the Roadmap.
## Definition for "Status Board"

I gave this its own ticket. See #167

Currently the document doesn't mention status boards but all of our discussions about pipelines assume that you have a place to view the project's Goals within a pipeline (ie. a Waffle Board). We should define the concept so we can use it explicitly in the model.

**Definition**: _Status Board_: A view of the project's Goals that shows their statuses within the project's pipeline.  This allows you to see the current state of work and to identify what work is ready to be picked up. 

Ie. You should be able to ask a PM "Where is this project's status board? What's the state of your project's board?" or a Project Lead can say "To see the current state of our Goals, look at the status board."

### LINKED ISSUE #167: Add Term: Status Board
refs #166 

Currently the document doesn't mention status boards but all of our discussions about pipelines assume that you have a place to view the project's Goals within a pipeline (ie. a Waffle Board). We should define the concept so we can use it explicitly in the model.

Definition: Status Board: A view of the project's Goals that shows their statuses within the project's pipeline. This allows you to see the current state of work and to identify what work is ready to be picked up.

Ie. You should be able to ask a PM "Where is this project's status board? What's the state of your project's board?" or a Project Lead can say "To see the current state of our Goals, look at the status board."

Note: This clearly belongs under the "Structures" heading of the PM Model doc as it is now but, as #168 explains, we might want to clarify those headings.

### LINKED ISSUE #168: Clarify the headings "Units of Work" and "Structures"
The PM Model document defines "Units of Work" and "Structures".
### Are these the right headings? Should we just define terms without headings?

What's the purpose of this section and these headings? Do the headings convey useful meaning that helps people use this model?
### Which of these are Artifacts? Which are just concepts or organizing principles?

The Units of work (Project, Milestone, Goal) are all artifacts -- they are things that exist in the project. It's a bit fuzzier with the "Structures" (Roadmap, Backlog and Pipeline). Based on discussions and examples, it seems like we expect the Roadmap to exist as an artifact, but Backlog and Pipeline are not explicit artifacts.  They're concepts or organizing principles.  This led to confusion [in these comments](https://github.com/ipfs/pm/pull/131/files/0fbf77efecf51695f4c9370c540e4d6727cda80d#r74582220)

I'm partially asking this question because I think we should add a term -- Status Board -- which is not an artifact. It's a dynamic view of the Goals & the Pipeline.  (see #167)

It does seem like a good idea for the Roadmap to be an explicit artifact -- ie. generated Roadmap.md generated by @haadcode's [roadmap generator](https://github.com/haadcode/roadmap-generator) because we can commit the Roadmap.md in into version control with the code, which lets us get the history of diffs applied to the Roadmap over time.

### LINKED ISSUE #169: Address "Definition of Ready" in the PM Document's discussion of Pipelines
See discussion in [comments on the PR](https://github.com/ipfs/pm/pull/131/files/0fbf77efecf51695f4c9370c540e4d6727cda80d#r74628251)

Add a note that a pipeline could distinguish between goals that are "Ready" or not ready based on some "Definition of Ready" -- ie. needing to be described in a way that is clear, testable, and feasible.

### PR DESCRIPTION
**Pull request by flyingzumwalt**
The commit messages are pretty self-explanatory.
- Completely Re-structured the document, headings and TOC for Readability (closes #168)
- Applies consistent use of terms like Backlog and Milestone with clear definitions (closes #165)
- Adds term "Status Board" (closes #167)
- Addresses "Definition of Ready" (closes #169)
- Frames Milestones as crossover point between the timeline view (Roadmap) and the pipeline view (Status Board) of your Goals (closes #166)
- Defines Project Lead, Product Owner, and (partially) Project Manager (ref #164)

### CODE REVIEW

**Comment by flyingzumwalt** [comment]
_**Question**: Should we just shelve this PR until the October 3rd sprint?_  A critical mass of the people who need to review this PR are crunching to prepare for Devcon (@diasdavid @whyrusleeping @haadcode @jbenet) and will need some time to decompress after devcon before they have headspace to look at this. 

If we return to this on the 3rd that will give us a week to merge the PR, make final edits and all give it a read-through before doing the Q4 Roadmap Updates in mid-October. I'm fine with that -- lets me focus on tutorials/primer, roadmaps, and helping with the devcon presentations/blog posts for the next 2 weeks.

**Comment by jbenet** [comment]
@flyingzumwalt shelving sgtm. i will get you some comments earlier than that, but i think most people are pretty heads down.

The one risk with shelving is: being on track with Q2 goals. do people feel behind on milestones?

**Comment by flyingzumwalt** [comment]
I think we've made huge improvements to the process informally. That has allowed people to get more work done and allowed them to keep a clearer view of their Roadmaps & Milestones. Pressing everyone to switch contexts into resuming the discussion about project management process before October will probably cause more harm than good in terms of hitting more Milestones.

File: drafts/project-management-process.md [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+
+## Artifacts
+
+### 📚 Projects
+
+_TODO: WTF is a Project? is a Project defined by its audience?_
+
+A **Project** is...
+
+_We **do not** use a one-to-one mapping between Projects and code bases._ Projects frequently include multiple code bases stored in multiple git repositories.
```
Comment (jbenet) [inline drafts/project-management-process.md]: Good to point out one project can involve many repos. but I think at least there will always be one repo that is the "root" of that project (eg `ipfs/go-ipfs` and `ipfs/js-ipfs` despite the many subrepos or unnamed subprojects).

**Comment by jbenet** [comment]
@flyingzumwalt this doc is very good, thank you for reworking it. It's very approachable, thorough, clear, and well-linked. I haven't found anything (beyond the clearly marked TODOs) to improve. 👍 👌 

I think maybe the "many roles" listing may be a bit intimidating to people coming in, and might seem "over organized". One thought: maybe the thing to emphasize with the word "role" is "hats" or "well-defined area of responsibility" instead of "well defined position". Some smaller projects will be run by one person most of the roles (wearing most of the hats), and some large ones will involve many people. I think role tends to be thought as "1 role per person", so maybe making it clear that it's "an area of responsibility", "a protocol actor", or something?

**Comment by haadcode** [comment]
This is excellent work @flyingzumwalt! You totally took the draft and made it awesome :)

One thing I would lose though: "Sprints". The term itself is loaded and we're not really doing sprints in the agile sense (and I think we shouldn't). I would keep to using the term Weekly Update as it frames the idea of non-sprint status update rather than something that follows the general idea of a sprint.

I would also try to re-word the "Methodology" section. It describes what we like, or what we take as an example and Scrum is definitely not one of them. Also, the bullet points list things that we haven't discussed nor agreed to, eg. "Embrace change by arranging work into sprints" or "Rely heavily on Test Driven Development". Especially the "Embrace change by arranging work into sprints" is imo something I wouldn't want to set as a guiding principle for our PM process.

Rest of is solid! It's so much nicer to read now.

File: drafts/project-management-process.md [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+A Milestone is completed when all Goals associated with it have been completed.
+
+**Attributes of a Milestone:**
+
+A Milestone has the following information:
+- Name
+  - Example 1: "Release go-ipfs v0.5.0"
+  - Example 2: "Working prototype of Pubsub implementation in go-ipfs"
+  - Example 3.1: "go-ipfs and js-ipfs implementations can interoperate"
+  - Example 3.2: "go-ipfs and js-ipfs clients use the same network"
```
Comment (diasdavid) [inline drafts/project-management-process.md]: I'm with @whyrusleeping and we are wondering if these `3.` milestones mean that there is a larger milestone composed by intermediate milestones or if it is just the same milestone written in 2 ways. @flyingzumwalt ?

Comment (flyingzumwalt) [inline drafts/project-management-process.md]: That particular section is directly from the original draft. @haadcode do you remember why the examples are written this way?

**Comment by flyingzumwalt** [comment]
@haadcode I added the bit about sprints because the term sprint is sprinkled across many IPFS pm documents and guides. For example, we call our weekly check-ins "sprint calls". I don't think we should use the term without defining it in this document. Are you suggesting that we should remove the term "sprint" from all of our documents, guides, calendars, etc?

Do you have a link to your presentation about Agile, Scrum and Kanban? I think we might have different understandings of what Scrum implies, so I want to update myself about your understanding of it.  I definitely see people doing stuff that I think of as Scrum habits and I personally draw ideas like "Definition of Ready" directly from Scrum.

The bullet list ("Strongly rely on TDD", etc.) is more of a placeholder. We should list the agile things that we do value but I'm not sure which things to include in that list or how to word them.

**Comment by diasdavid** [comment]
One important thing to add in consideration, was that with this last Q4 meeting, Milestones leads were introduced, which can be one or more per milestone, they have the responsibility of owning the milestone, while the Captain ensures they are not blocked to do so.

**Comment by flyingzumwalt** [comment]
I'd like to merge this PR now. How many 👍's do I need to get in order for someone to merge it?

cc @jbenet @haadcode @whyrusleeping @diasdavid @RichardLitt

**Comment by flyingzumwalt** [comment]
(Whoever merges this, I recommend squashing the commits when you merge)

**Comment by RichardLitt** [comment]
You just need one. If you think it is ready, it is ready.

**Pull request merged**